File: covilo-core/src/main/java/me/kvdpxne/covilo/presentation/dto/UserDto.java
package me.kvdpxne.covilo.presentation.dto;

import java.time.LocalDate;
import java.time.LocalDateTime;
import me.kvdpxne.covilo.domain.model.Gender;
import me.kvdpxne.covilo.domain.model.User;

/**
 *
 */
public record UserDto(
  String identifier,
  String email,
  String firstName,
  String lastName,
  Gender gender,
  LocalDate birthDate,
  LocalDateTime createdDate,
  LocalDateTime lastModifiedDate
) {

  /**
   *
   */
  public static UserDto from(
    final User user
  ) {
    return new UserDto(
      user.getIdentifier(),
      user.getLowerName(),
      user.getFirstName(),
      user.getLastName(),
      user.getGender(),
      user.getBirthDate(),
      user.getCreatedDate(),
      user.getLastModifiedDate()
    );
  }
}

File: covilo-api/src/main/kotlin/me/kvdpxne/covilo/rest/api/dto/UserDto.kt
package me.kvdpxne.covilo.rest.api.dto

import me.kvdpxne.covilo.domain.models.User
import me.kvdpxne.covilo.domain.models.Users

/**
 * Shortcut to the [UserDto] collection.
 */
typealias UsersDto = Collection<UserDto>

/**
 * TODO doc
 */
data class UserDto(
  val identifier: String,
  val email: String,
  val firstName: String,
  val lastName: String,
  val brithDate: String,
  val createdDate: String,
  val lastModifiedDate: String
)

/**
 * Transforms the [User] entity to [UserDto].
 */
fun User.toDto(): UserDto {
  return UserDto(
    identifier.toString(),
    email,
    firstName.toString(),
    lastName.toString(),
    birthDate.toString(),
    createdDate.toString(),
    lastModifiedDate.toString()
  )
}

/**
 * Transforms all [User] entities that are in the collection
 * and are not null to [UserDto].
 */
fun Users.toDto(): UsersDto {
  return mapNotNull {
    it.toDto()
  }
}